README.md
# NAME

Plack::Middleware::Antibot - Prevent bots from submitting forms

# SYNOPSIS

    use Plack::Builder;

    my $app = { ... };

    builder {
        enable 'Antibot', filters => [qw/FakeField TooFast/];
        $app;
    };

# DESCRIPTION

Plack::Middleware::Antibot is a [Plack](https://metacpan.org/pod/Plack) middleware that prevents bots from
submitting forms. Every filter implements its own checks, so see their
documentation.

Plack::Middleware::Antibot uses scoring system (0 to 1) to determine if the
client is a bot. Thus it can be configured to match any needs.

## `$env`

Some filters set additional `$env` keys all prefixed with `antibot.`. For
example `TextCaptcha` filter sets `antibot.text_captcha` to be shown to the
user.

## Options

### **max\_score**

When accumulated score reaches this amount, no more filters are run and bot is
detected. `0.8` by default.

### **filters**

    enable 'Antibot', filters => ['FakeField'];

To specify filter arguments instead of a filter name pass an array references:

    enable 'Antibot', filters => [['FakeField', field_name => 'my_fake_field']];

### **fall\_through**

    enable 'Antibot', filters => ['FakeField'], fall_through => 1;

Sometimes it is needed to process detected bot yourself. This way in case of
detection `$env`'s key `antibot.detected` will be set.

## Available filters

- [Plack::Middleware::Antibot::FakeField](https://metacpan.org/pod/Plack::Middleware::Antibot::FakeField) (requires [Plack::Session](https://metacpan.org/pod/Plack::Session))

    Check if an invisible or hidden field is submitted.

- [Plack::Middleware::Antibot::Static](https://metacpan.org/pod/Plack::Middleware::Antibot::Static) (requires [Plack::Session](https://metacpan.org/pod/Plack::Session))

    Check if a static file was fetched before form submission.

- [Plack::Middleware::Antibot::TextCaptcha](https://metacpan.org/pod/Plack::Middleware::Antibot::TextCaptcha) (requires [Plack::Session](https://metacpan.org/pod/Plack::Session))

    Check if correct random text captcha is submitted.

- [Plack::Middleware::Antibot::TooFast](https://metacpan.org/pod/Plack::Middleware::Antibot::TooFast)

    Check if form is submitted too fast.

- [Plack::Middleware::Antibot::TooSlow](https://metacpan.org/pod/Plack::Middleware::Antibot::TooSlow)

    Check if form is submitted too slow.

# LICENSE

Copyright (C) vti.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

vti <viacheslav.t@gmail.com>


Changes
Revision history for Perl extension Plack-Middleware-Antibot

{{$NEXT}}

0.02 2015-02-05T07:26:40Z

    - specify dependencies

0.01 2015-02-04T14:57:31Z

    - original version



lib/Plack/Middleware/Antibot/Static.pm
package Plack::Middleware::Antibot::Static;

use strict;
use warnings;

use parent 'Plack::Middleware::Antibot::FilterBase';

use Plack::Session;

sub new {
    my $self = shift->SUPER::new(@_);
    my (%params) = @_;

    $self->{path}         = $params{path}         || '/antibot.gif';
    $self->{session_name} = $params{session_name} || 'antibot_static';
    $self->{timeout}      = $params{timeout}      || 60 * 15;
    $self->{score}        = $params{score}        || 0.9;

    return $self;
}

sub execute {
    my $self = shift;
    my ($env) = @_;

    if ($env->{REQUEST_METHOD} eq 'GET') {
        my $path_info = $env->{PATH_INFO};

        if ($path_info eq $self->{path}) {
            my $session = Plack::Session->new($env);
            $session->set($self->{session_name} => time);

            return [200, [], ['']];
        }

        $env->{'plack.antibot.static.path'} = $self->{path};
        $env->{'plack.antibot.static.html'} = qq{<img src="$self->{path}" }
          . qq{width="1" height="1" style="display:none" />};
    }
    elsif ($env->{REQUEST_METHOD} eq 'POST') {
        my $session = Plack::Session->new($env);

        my $static_fetched = $session->get($self->{session_name});
        unless ($static_fetched && time - $static_fetched < $self->{timeout}) {
            $env->{'plack.antibot.static.detected'}++;
        }
    }

    return;
}

1;
__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::Antibot::Static - Check if static file was fetched

=head1 SYNOPSIS

    enable 'Antibot', filters =>
      [['Static', path => '/antibot.css']];

=head1 DESCRIPTION

Plack::Middleware::Antibot::Static checks if a static-like file was fetched.

=head2 Options

=head3 B<score>

Filter's score when bot detected. C<0.9> by default.

=head3 B<session_name>

Session name. C<antibot_static> by default.

=head3 B<timeout>

Expiration timeout in seconds. C<15 * 60> by default (15 minutes).

=cut


lib/Plack/Middleware/Antibot/TooFast.pm
package Plack::Middleware::Antibot::TooFast;

use strict;
use warnings;

use parent 'Plack::Middleware::Antibot::FilterBase';

use Plack::Session;

sub new {
    my $self = shift->SUPER::new(@_);
    my (%params) = @_;

    $self->{session_name} = $params{session_name} || 'antibot_toofast';
    $self->{timeout}      = $params{timeout}      || 1;
    $self->{score}        = $params{score}        || 0.9;

    return $self;
}

sub execute {
    my $self = shift;
    my ($env) = @_;

    if ($env->{REQUEST_METHOD} eq 'GET') {
        my $session = Plack::Session->new($env);

        $session->set($self->{session_name}, time);
    }
    elsif ($env->{REQUEST_METHOD} eq 'POST') {
        my $session = Plack::Session->new($env);

        my $too_fast = $session->get($self->{session_name});
        unless ($too_fast && time - $too_fast > $self->{timeout}) {
            $env->{'plack.antibot.toofast.detected'}++;
        }
    }

    return;
}

1;
__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::Antibot::TooFast - Check if form was submitted too fast

=head1 SYNOPSIS

    enable 'Antibot', filters => ['TooFast'];

=head1 DESCRIPTION

Plack::Middleware::Antibot::TooFast checks if form was submitted too fast.

=head2 Options

=head3 B<score>

Filter's score when bot detected. C<0.8> by default.

=head3 B<session_name>

Session name. C<antibot_toofast> by default.

=head3 B<timeout>

Timeout in seconds. C<1> by default.

=cut


lib/Plack/Middleware/Antibot/TooSlow.pm
package Plack::Middleware::Antibot::TooSlow;

use strict;
use warnings;

use parent 'Plack::Middleware::Antibot::FilterBase';

use Plack::Session;

sub new {
    my $self = shift->SUPER::new(@_);
    my (%params) = @_;

    $self->{session_name} = $params{session_name} || 'antibot_tooslow';
    $self->{timeout}      = $params{timeout}      || 60 * 60;
    $self->{score}        = $params{score}        || 0.8;

    return $self;
}

sub execute {
    my $self = shift;
    my ($env) = @_;

    if ($env->{REQUEST_METHOD} eq 'GET') {
        my $session = Plack::Session->new($env);

        $session->set($self->{session_name}, time);
    }
    elsif ($env->{REQUEST_METHOD} eq 'POST') {
        my $session = Plack::Session->new($env);

        my $too_slow = $session->get($self->{session_name});
        unless ($too_slow && time - $too_slow < $self->{timeout}) {
            $env->{'plack.antibot.tooslow.detected'}++;
        }
    }

    return;
}

1;
__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::Antibot::TooSlow - Check if form was submitted too slow

=head1 SYNOPSIS

    enable 'Antibot', filters => ['TooSlow'];

=head1 DESCRIPTION

Plack::Middleware::Antibot::TooSlow checks if form was submitted too slow.

=head2 Options

=head3 B<score>

Filter's score when bot detected. C<0.8> by default.

=head3 B<session_name>

Session name. C<antibot_tooslow> by default.

=head3 B<timeout>

Timeout in seconds. C<60 * 60> by default (1 hour).

=cut


lib/Plack/Middleware/Antibot/FakeField.pm
package Plack::Middleware::Antibot::FakeField;

use strict;
use warnings;

use parent 'Plack::Middleware::Antibot::FilterBase';

use Plack::Request;

sub new {
    my $self = shift->SUPER::new(@_);
    my (%params) = @_;

    $self->{field_name} = $params{field_name} || 'antibot_fake_field';
    $self->{score}      = $params{score}      || 0.8;

    return $self;
}

sub execute {
    my $self = shift;
    my ($env) = @_;

    if ($env->{REQUEST_METHOD} eq 'POST') {
        my $value = Plack::Request->new($env)->param($self->{field_name});
        if (defined $value && length $value) {
            $env->{'plack.antibot.fakefield.detected'}++;
        }
    }
    else {
        $env->{'plack.antibot.fakefield.field_name'} = $self->{field_name};
        $env->{'plack.antibot.fakefield.html'} = <<"EOF";
<div style="display:none">
<label>Please leave blank</label>
<input name="$self->{field_name}" />
</div>
EOF
    }

    return;
}

1;
__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::Antibot::FakeField - Check if fake field was submitted

=head1 SYNOPSIS

    enable 'Antibot', filters => ['FakeField'];

=head1 DESCRIPTION

Plack::Middleware::Antibot::FakeField checks if a fake field was submitted. The
field with specified name has to be present on a form, but should be invisible
to user. This can be achieved either by CSS or by JavaScript.

    <div style="display:none">
        <label>Please leave this blank</label>
        <input name="antibot_fake_field" />
    </div>

It is better to name the field to something that makes sense but doesn't clash
with other fields.

=head2 Options

=head3 B<score>

Filter's score when bot detected. C<0.8> by default.

=head3 B<field_name>

Field name. C<antibot_fake_field> by default.

=cut


lib/Plack/Middleware/Antibot/TextCaptcha.pm
package Plack::Middleware::Antibot::TextCaptcha;

use strict;
use warnings;

use parent 'Plack::Middleware::Antibot::FilterBase';

use Plack::Session;
use Plack::Request;

sub new {
    my $self = shift->SUPER::new(@_);
    my (%params) = @_;

    $self->{session_name} = $params{session_name} || 'antibot_textcaptcha';
    $self->{field_name}   = $params{field_name}   || 'antibot_textcaptcha';
    $self->{variants} = $params{variants} || [{text => '2 + 2', answer => 4}];
    $self->{score}    = $params{score}    || 0.9;

    return $self;
}

sub execute {
    my $self = shift;
    my ($env) = @_;

    my $variants = $self->{variants};

    my $captcha = $variants->[int(rand(@$variants))];
    $env->{'plack.antibot.textcaptcha.text'}       = $captcha->{text};
    $env->{'plack.antibot.textcaptcha.field_name'} = $self->{field_name};

    if ($env->{REQUEST_METHOD} eq 'GET') {
        my $session = Plack::Session->new($env);
        $session->set($self->{session_name}, $captcha->{answer});
    }
    elsif ($env->{REQUEST_METHOD} eq 'POST') {
        my $session = Plack::Session->new($env);

        my $expected = $session->get($self->{session_name});
        my $got      = Plack::Request->new($env)->param($self->{field_name});

        unless ($expected && $got && $got eq $expected) {
            $env->{'plack.antibot.textcaptcha.detected'}++;
        }

        $session->set($self->{session_name}, $captcha->{answer});
    }

    return;
}

1;
__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::Antibot::TextCaptcha - Check if correct captcha was submitted

=head1 SYNOPSIS

    enable 'Antibot', filters =>
      [['TextCaptcha', variants => [{text => '2 + 2', answer => 4}]]];

=head1 DESCRIPTION

Plack::Middleware::Antibot::TextCaptcha checks if a correct captcha was
submitted. Most of the time a simple text with a simple solution is enough to
prevent bots from successful form submitions.

=head2 C<$env>

This filter sets C<antibot.text_captcha> as captcha text. This should be shown
to the user as a field label or description.

=head2 Options

=head3 B<score>

Filter's score when bot detected. C<0.8> by default.

=head3 B<session_name>

Session name. C<antibot_textcaptcha> by default.

=head3 B<field_name>

Field name. C<antibot_textcaptcha> by default.

=head3 B<variants>

Captcha variants. C<[{text => '2 + 2', answer => 4}]> by default.

=cut


lib/Plack/Middleware/Antibot.pm
package Plack::Middleware::Antibot;

use strict;
use warnings;

use parent 'Plack::Middleware';

our $VERSION = "0.01";

use List::Util qw(sum reduce);
use Plack::Util ();
use Plack::Util::Accessor qw(filters fall_through max_score);

sub prepare_app {
    my $self = shift;

    $self->{max_score} ||= 0.8;

    my $filters_names = $self->filters;

    my @filters;
    foreach my $filter (@$filters_names) {
        my @args;
        if (ref $filter eq 'ARRAY') {
            my $ref = $filter;
            $filter = shift @$ref;
            @args   = @$ref;
        }

        my $filter_class = __PACKAGE__ . '::' . $filter;

        Plack::Util::load_class($filter_class);

        push @filters, $filter_class->new(@args);
    }

    $self->filters(\@filters);

    return $self;
}

sub call {
    my $self = shift;
    my ($env) = @_;

    my @scores;
    my $current_score = 0;
    foreach my $filter (@{$self->filters}) {
        my $res = $filter->execute($env);
        return $res if $res && ref $res eq 'ARRAY';

        my $name = (split /::/, ref $filter)[-1];
        my $key = 'plack.antibot.' . lc($name) . '.detected';

        if ($env->{$key}) {
            push @scores, $filter->score;

            if (@scores > 1) {
                my $p = sum @scores;
                my $q = reduce { $a * $b } @scores;

                $current_score = $p - $q;
            }
            else {
                $current_score = $filter->score;
            }
        }

        last if $current_score >= $self->max_score;
    }

    $env->{'plack.antibot.score'} = $current_score;

    if ($current_score >= $self->max_score) {
        $env->{'plack.antibot.detected'} = 1;

        return [400, [], ['Bad request']] unless $self->fall_through;
    }

    return $self->app->($env);
}

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Plack::Middleware::Antibot - Prevent bots from submitting forms

=head1 SYNOPSIS

    use Plack::Builder;

    my $app = { ... };

    builder {
        enable 'Antibot', filters => [qw/FakeField TooFast/];
        $app;
    };

=head1 DESCRIPTION

Plack::Middleware::Antibot is a L<Plack> middleware that prevents bots from
submitting forms. Every filter implements its own checks, so see their
documentation.

Plack::Middleware::Antibot uses scoring system (0 to 1) to determine if the
client is a bot. Thus it can be configured to match any needs.

=head2 C<$env>

Some filters set additional C<$env> keys all prefixed with C<antibot.>. For
example C<TextCaptcha> filter sets C<antibot.text_captcha> to be shown to the
user.

=head2 Options

=head3 B<max_score>

When accumulated score reaches this amount, no more filters are run and bot is
detected. C<0.8> by default.

=head3 B<filters>

    enable 'Antibot', filters => ['FakeField'];

To specify filter arguments instead of a filter name pass an array references:

    enable 'Antibot', filters => [['FakeField', field_name => 'my_fake_field']];

=head3 B<fall_through>

    enable 'Antibot', filters => ['FakeField'], fall_through => 1;

Sometimes it is needed to process detected bot yourself. This way in case of
detection C<$env>'s key C<antibot.detected> will be set.

=head2 Available filters

=over

=item L<Plack::Middleware::Antibot::FakeField> (requires L<Plack::Session>)

Check if an invisible or hidden field is submitted.

=item L<Plack::Middleware::Antibot::Static> (requires L<Plack::Session>)

Check if a static file was fetched before form submission.

=item L<Plack::Middleware::Antibot::TextCaptcha> (requires L<Plack::Session>)

Check if correct random text captcha is submitted.

=item L<Plack::Middleware::Antibot::TooFast>

Check if form is submitted too fast.

=item L<Plack::Middleware::Antibot::TooSlow>

Check if form is submitted too slow.

=back

=head1 ISA

L<Plack::Middleware>

=head1 METHODS

=head2 C<prepare_app>

=head2 C<call($env)>

=head1 INHERITED METHODS

=head2 C<wrap($app, @args)>

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut
